File: uwp/ArtistWayUWP/Services/SyncService.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Windows.Data.Json;

namespace ArtistWayUWP.Services
{
    // Sincroniza os "stores" do LocalDataStore com o Firestore: puxa a
    // versão da nuvem, mescla com a local registro a registro por
    // updatedAt (quem for mais recente vence), grava o resultado local e
    // sobe de volta -- sempre nos dois sentidos, sempre idempotente. Sem
    // listener em tempo real: cada chamada é só um request/response HTTP
    // curto (ver "Decisões de arquitetura" em sincronizacao-nuvem-setup.md).
    public static class SyncService
    {
        private const string ProjectId = "theartistsway";

        public static async Task<string> SyncAllAsync()
        {
            FirebaseSession session = SessionService.GetSession();
            if (session == null)
            {
                return "Não logado -- nada pra sincronizar.";
            }

            string idToken = session.IdToken;
            if (session.NeedsRefresh)
            {
                idToken = await RefreshIdTokenAsync(session);
                if (idToken == null)
                {
                    return "Sessão expirada -- entre de novo.";
                }
            }

            try
            {
                using (HttpClient client = new HttpClient())
                {
                    client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", idToken);
                    foreach (string storeName in LocalDataStore.SyncStoreNames)
                    {
                        await SyncStoreAsync(client, session.Uid, storeName);
                    }
                }
                return "Sincronizado às " + DateTimeOffset.Now.ToString("HH:mm");
            }
            catch (Exception ex)
            {
                return "Falha ao sincronizar (tentará de novo mais tarde): " + ex.Message;
            }
        }

        private static async Task SyncStoreAsync(HttpClient client, string uid, string storeName)
        {
            JsonObject local = await LocalDataStore.GetStoreForSyncAsync(storeName);
            JsonObject remote = await GetRemoteStoreAsync(client, uid, storeName);

            JsonObject merged = storeName == "settings"
                ? MergeWholeBlob(local, remote)
                : MergePerRecord(local, remote, TimestampFieldFor(storeName));

            await LocalDataStore.WriteStoreForSyncAsync(storeName, merged);
            await PutRemoteStoreAsync(client, uid, storeName, merged);
        }

        private static string TimestampFieldFor(string storeName) => storeName == "checkins" ? "savedAt" : "updatedAt";

        // ---------- mesclagem ----------

        private static JsonObject MergeWholeBlob(JsonObject local, JsonObject remote)
        {
            if (remote.Count == 0)
            {
                return local;
            }
            if (local.Count == 0)
            {
                return remote;
            }
            DateTimeOffset localTs = ParseTimestamp(local, "_updatedAt");
            DateTimeOffset remoteTs = ParseTimestamp(remote, "_updatedAt");
            return remoteTs > localTs ? remote : local;
        }

        private static JsonObject MergePerRecord(JsonObject local, JsonObject remote, string tsField)
        {
            JsonObject merged = new JsonObject();
            HashSet<string> keys = new HashSet<string>();
            foreach (KeyValuePair<string, IJsonValue> kv in local)
            {
                keys.Add(kv.Key);
            }
            foreach (KeyValuePair<string, IJsonValue> kv in remote)
            {
                keys.Add(kv.Key);
            }

            foreach (string key in keys)
            {
                bool hasLocal = local.ContainsKey(key);
                bool hasRemote = remote.ContainsKey(key);
                if (hasLocal && hasRemote)
                {
                    JsonObject localEntry = NormalizeRecord(local[key]);
                    JsonObject remoteEntry = NormalizeRecord(remote[key]);
                    DateTimeOffset localTs = ParseTimestamp(localEntry, tsField);
                    DateTimeOffset remoteTs = ParseTimestamp(remoteEntry, tsField);
                    merged[key] = remoteTs > localTs ? remoteEntry : localEntry;
                }
                else if (hasLocal)
                {
                    merged[key] = NormalizeRecord(local[key]);
                }
                else
                {
                    merged[key] = NormalizeRecord(remote[key]);
                }
            }
            return merged;
        }

        // Registros de antes da sincronização existir vinham como booleano
        // solto (sem updatedAt) -- normaliza pro formato novo, tratado como
        // "o mais antigo possível" na comparação (ParseTimestamp devolve
        // DateTimeOffset.MinValue quando falta o campo).
        private static JsonObject NormalizeRecord(IJsonValue value)
        {
            if (value.ValueType == JsonValueType.Object)
            {
                return value.GetObject();
            }
            JsonObject wrapped = new JsonObject();
            if (value.ValueType == JsonValueType.Boolean)
            {
                wrapped["done"] = JsonValue.CreateBooleanValue(value.GetBoolean());
            }
            return wrapped;
        }

        private static DateTimeOffset ParseTimestamp(JsonObject obj, string field)
        {
            if (obj.ContainsKey(field) && obj[field].ValueType == JsonValueType.String &&
                DateTimeOffset.TryParse(obj[field].GetString(), out DateTimeOffset result))
            {
                return result;
            }
            return DateTimeOffset.MinValue;
        }

        // ---------- Firestore REST ----------
        // Cada store vira um documento com um único campo "data" (o JSON
        // inteiro do store, como string) -- evita ter que traduzir pro
        // formato de tipos nativos do Firestore (mapValue/arrayValue/etc.)
        // pra estruturas que, na prática, o app só lê/escreve como blob.

        private static string DocUrl(string uid, string storeName) =>
            $"https://firestore.googleapis.com/v1/projects/{ProjectId}/databases/(default)/documents/users/{uid}/stores/{storeName}";

        private static async Task<JsonObject> GetRemoteStoreAsync(HttpClient client, string uid, string storeName)
        {
            HttpResponseMessage response = await client.GetAsync(DocUrl(uid, storeName));
            if (response.StatusCode == HttpStatusCode.NotFound)
            {
                return new JsonObject();
            }
            if (!response.IsSuccessStatusCode)
            {
                throw new Exception($"Firestore GET {storeName}: {(int)response.StatusCode}");
            }

            string text = await response.Content.ReadAsStringAsync();
            JsonObject doc = JsonObject.Parse(text);
            if (!doc.ContainsKey("fields"))
            {
                return new JsonObject();
            }
            JsonObject fields = doc["fields"].GetObject();
            if (!fields.ContainsKey("data"))
            {
                return new JsonObject();
            }
            string dataJson = fields["data"].GetObject()["stringValue"].GetString();
            return JsonObject.Parse(dataJson);
        }

        private static async Task PutRemoteStoreAsync(HttpClient client, string uid, string storeName, JsonObject data)
        {
            JsonObject body = new JsonObject
            {
                ["fields"] = new JsonObject
                {
                    ["data"] = new JsonObject { ["stringValue"] = JsonValue.CreateStringValue(data.Stringify()) },
                    ["updatedAt"] = new JsonObject { ["timestampValue"] = JsonValue.CreateStringValue(DateTimeOffset.UtcNow.ToString("o")) },
                },
            };

            HttpContent content = new StringContent(body.Stringify(), Encoding.UTF8, "application/json");
            HttpRequestMessage request = new HttpRequestMessage(new HttpMethod("PATCH"), DocUrl(uid, storeName)) { Content = content };
            HttpResponseMessage response = await client.SendAsync(request);
            if (!response.IsSuccessStatusCode)
            {
                string errorText = await response.Content.ReadAsStringAsync();
                throw new Exception($"Firestore PATCH {storeName}: {(int)response.StatusCode} {errorText}");
            }
        }

        // ---------- renovação de token ----------

        private static async Task<string> RefreshIdTokenAsync(FirebaseSession session)
        {
            try
            {
                using (HttpClient client = new HttpClient())
                {
                    FormUrlEncodedContent body = new FormUrlEncodedContent(new Dictionary<string, string>
                    {
                        ["grant_type"] = "refresh_token",
                        ["refresh_token"] = session.RefreshToken,
                    });
                    HttpResponseMessage response = await client.PostAsync(
                        $"https://securetoken.googleapis.com/v1/token?key={AuthService.FirebaseApiKey}", body);
                    string text = await response.Content.ReadAsStringAsync();
                    if (!response.IsSuccessStatusCode)
                    {
                        return null;
                    }

                    JsonObject json = JsonObject.Parse(text);
                    string idToken = json.ContainsKey("id_token") ? json["id_token"].GetString() : null;
                    int expiresIn = json.ContainsKey("expires_in") && int.TryParse(json["expires_in"].GetString(), out int e) ? e : 3600;
                    if (string.IsNullOrEmpty(idToken))
                    {
                        return null;
                    }
                    SessionService.UpdateTokens(idToken, expiresIn);
                    return idToken;
                }
            }
            catch (Exception)
            {
                return null;
            }
        }
    }
}

File: android/ArtistWayAndroid/app/src/main/java/com/rodcarvalho/artistway/sync/SyncService.kt
package com.rodcarvalho.artistway.sync

import com.rodcarvalho.artistway.auth.FirebaseConfig
import com.rodcarvalho.artistway.auth.FirebaseSession
import com.rodcarvalho.artistway.auth.SessionService
import com.rodcarvalho.artistway.data.LocalDataStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.time.Instant
import java.time.LocalTime

// Sincroniza os "stores" do LocalDataStore com o Firestore: puxa a
// versão da nuvem, mescla com a local registro a registro por
// updatedAt (quem for mais recente vence), grava o resultado local e
// sobe de volta — sempre nos dois sentidos, sempre idempotente. Porta
// quase linha a linha de SyncService.cs (UWP); HttpURLConnection puro
// em vez de uma lib de rede nova só pra essas 4 chamadas.
object SyncService {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun syncAll(): String = withContext(Dispatchers.IO) {
        val session = SessionService.getSession() ?: return@withContext "Não logado — nada pra sincronizar."
        val idToken = if (session.needsRefresh()) {
            refreshIdToken(session) ?: return@withContext "Sessão expirada — entre de novo."
        } else {
            session.idToken
        }

        try {
            for (storeName in LocalDataStore.SYNC_STORE_NAMES) {
                syncStore(idToken, session.uid, storeName)
            }
            "Sincronizado às ${LocalTime.now().withNano(0)}"
        } catch (e: Exception) {
            "Falha ao sincronizar (tentará de novo mais tarde): ${e.message}"
        }
    }

    // Apaga os dados da nuvem (todos os stores) sem mexer no login —
    // usado pelo reset "Apagar meus dados" (mantém o aparelho logado).
    suspend fun clearCloudData(): Boolean = withContext(Dispatchers.IO) {
        val session = SessionService.getSession() ?: return@withContext true
        val idToken = if (session.needsRefresh()) {
            refreshIdToken(session) ?: return@withContext false
        } else {
            session.idToken
        }

        try {
            for (storeName in LocalDataStore.SYNC_STORE_NAMES) {
                val conn = openConnection(docUrl(session.uid, storeName), "DELETE", idToken)
                val code = conn.responseCode
                conn.disconnect()
                if (code !in 200..299 && code != 404) return@withContext false
            }
            true
        } catch (e: Exception) {
            false
        }
    }

    private suspend fun syncStore(idToken: String, uid: String, storeName: String) {
        val local = LocalDataStore.getStoreForSync(storeName)
        val remote = getRemoteStore(idToken, uid, storeName)
        val merged = if (storeName == "settings") {
            mergeWholeBlob(local, remote)
        } else {
            mergePerRecord(local, remote, if (storeName == "checkins") "savedAt" else "updatedAt")
        }
        LocalDataStore.writeStoreForSync(storeName, merged)
        putRemoteStore(idToken, uid, storeName, merged)
    }

    // — mesclagem —

    private fun mergeWholeBlob(local: JsonObject, remote: JsonObject): JsonObject {
        if (remote.isEmpty()) return local
        if (local.isEmpty()) return remote
        val localTs = parseTimestamp(local, "_updatedAt")
        val remoteTs = parseTimestamp(remote, "_updatedAt")
        return if (remoteTs.isAfter(localTs)) remote else local
    }

    private fun mergePerRecord(local: JsonObject, remote: JsonObject, tsField: String): JsonObject {
        val keys = LinkedHashSet<String>().apply { addAll(local.keys); addAll(remote.keys) }
        val merged = LinkedHashMap<String, JsonElement>()
        for (key in keys) {
            val hasLocal = local.containsKey(key)
            val hasRemote = remote.containsKey(key)
            merged[key] = when {
                hasLocal && hasRemote -> {
                    val localEntry = normalizeRecord(local.getValue(key))
                    val remoteEntry = normalizeRecord(remote.getValue(key))
                    val localTs = parseTimestamp(localEntry, tsField)
                    val remoteTs = parseTimestamp(remoteEntry, tsField)
                    if (remoteTs.isAfter(localTs)) remoteEntry else localEntry
                }
                hasLocal -> normalizeRecord(local.getValue(key))
                else -> normalizeRecord(remote.getValue(key))
            }
        }
        return JsonObject(merged)
    }

    // Registros de antes da sincronização existir vinham como booleano
    // solto (sem updatedAt) — normaliza pro formato novo, tratado como
    // "o mais antigo possível" na comparação.
    private fun normalizeRecord(value: JsonElement): JsonObject {
        if (value is JsonObject) return value
        val prim = value as? JsonPrimitive
        val boolValue = prim?.booleanOrNull
        return if (boolValue != null) buildJsonObject { put("done", boolValue) } else JsonObject(emptyMap())
    }

    private fun parseTimestamp(obj: JsonObject, field: String): Instant {
        val text = (obj[field] as? JsonPrimitive)?.takeIf { it.isString }?.content ?: return Instant.MIN
        return runCatching { Instant.parse(text) }.getOrDefault(Instant.MIN)
    }

    // — Firestore REST —
    // Cada store vira um documento com um único campo "data" (o JSON
    // inteiro do store, como string) — evita ter que traduzir pro
    // formato de tipos nativos do Firestore pra estruturas que, na
    // prática, o app só lê/escreve como blob.

    private fun docUrl(uid: String, storeName: String) =
        "https://firestore.googleapis.com/v1/projects/${FirebaseConfig.PROJECT_ID}/databases/(default)/documents/users/$uid/stores/$storeName"

    private fun getRemoteStore(idToken: String, uid: String, storeName: String): JsonObject {
        val conn = openConnection(docUrl(uid, storeName), "GET", idToken)
        val code = conn.responseCode
        if (code == 404) {
            conn.disconnect()
            return JsonObject(emptyMap())
        }
        if (code !in 200..299) {
            conn.disconnect()
            throw Exception("Firestore GET $storeName: $code")
        }
        val text = conn.inputStream.bufferedReader().use { it.readText() }
        conn.disconnect()
        val doc = json.parseToJsonElement(text).jsonObject
        val fields = doc["fields"] as? JsonObject ?: return JsonObject(emptyMap())
        val dataField = fields["data"] as? JsonObject ?: return JsonObject(emptyMap())
        val dataJson = (dataField["stringValue"] as? JsonPrimitive)?.content ?: return JsonObject(emptyMap())
        return json.parseToJsonElement(dataJson).jsonObject
    }

    private fun putRemoteStore(idToken: String, uid: String, storeName: String, data: JsonObject) {
        val body = buildJsonObject {
            put(
                "fields",
                buildJsonObject {
                    put("data", buildJsonObject { put("stringValue", data.toString()) })
                    put("updatedAt", buildJsonObject { put("timestampValue", Instant.now().toString()) })
                },
            )
        }
        val conn = openConnection(docUrl(uid, storeName), "PATCH", idToken)
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
        OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(body.toString()) }
        val code = conn.responseCode
        if (code !in 200..299) {
            val errorText = (conn.errorStream ?: conn.inputStream)?.bufferedReader()?.use { it.readText() }
            conn.disconnect()
            throw Exception("Firestore PATCH $storeName: $code $errorText")
        }
        conn.disconnect()
    }

    private fun openConnection(url: String, method: String, idToken: String): HttpURLConnection {
        val conn = URL(url).openConnection() as HttpURLConnection
        conn.requestMethod = method
        conn.setRequestProperty("Authorization", "Bearer $idToken")
        conn.connectTimeout = 15000
        conn.readTimeout = 15000
        return conn
    }

    // — renovação de token —

    private fun refreshIdToken(session: FirebaseSession): String? {
        return try {
            val body = "grant_type=refresh_token&refresh_token=" + URLEncoder.encode(session.refreshToken, "UTF-8")
            val conn = URL("https://securetoken.googleapis.com/v1/token?key=${FirebaseConfig.API_KEY}").openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.connectTimeout = 15000
            conn.readTimeout = 15000
            conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(body) }
            val code = conn.responseCode
            val text = (if (code in 200..299) conn.inputStream else conn.errorStream)?.bufferedReader()?.use { it.readText() }
            conn.disconnect()
            if (code !in 200..299 || text == null) return null

            val obj = json.parseToJsonElement(text).jsonObject
            val idToken = (obj["id_token"] as? JsonPrimitive)?.content ?: return null
            val expiresIn = (obj["expires_in"] as? JsonPrimitive)?.content?.toIntOrNull() ?: 3600
            SessionService.updateTokens(idToken, expiresIn)
            idToken
        } catch (e: Exception) {
            null
        }
    }
}
